browser/ui/webui/ai_chat/ai_chat_ui_page_handler.cc

```diff
@@ -444,4 +564,122 @@ void AIChatUIPageHandler::OnGetPremiumStatus(
   }
 }
 
+#if BUILDFLAG(ENABLE_PRINT_PREVIEW)
+void AIChatUIPageHandler::PreviewCleanup() {
+  auto preview_ui_id = owner_->GetPreviewUIId();
+  CHECK(preview_ui_id);
+  PrintPreviewDataService::GetInstance()->RemoveEntry(*preview_ui_id);
+  print_render_frame_->OnPrintPreviewDialogClosed();
+  owner_->DisconnectPrintPrieviewUI();
+}
+
+void AIChatUIPageHandler::OnPreviewReady() {
+  scoped_refptr<base::RefCountedMemory> data;
+  auto preview_ui_id = owner_->GetPreviewUIId();
+  CHECK(preview_ui_id);
+  PrintPreviewDataService::GetInstance()->GetDataEntry(
+      *preview_ui_id, printing::COMPLETE_PREVIEW_DOCUMENT_INDEX, &data);
+  if (!data.get()) {
+    DLOG(ERROR) << "no data from preview id: " << *preview_ui_id;
+    return;
+  }
+  auto pdf_region = base::ReadOnlySharedMemoryRegion::Create(data->size());
+  if (!pdf_region.IsValid()) {
+    DLOG(ERROR) << "Failed allocate memory for PDF file";
+    return;
+  }
+  memcpy(pdf_region.mapping.memory(), data->data(), data->size());
+  std::optional<bool> pdf_use_skia_renderer_enabled;
+  auto* prefs = profile_->GetPrefs();
+  if (prefs &&
+      prefs->IsManagedPreference(::prefs::kPdfUseSkiaRendererEnabled)) {
+    pdf_use_skia_renderer_enabled =
+        prefs->GetBoolean(::prefs::kPdfUseSkiaRendererEnabled);
+  }
+  preview_page_text_extractor_ = std::make_unique<PreviewPageTextExtractor>(
+      std::move(pdf_region.region),
+      base::BindOnce(&AIChatUIPageHandler::OnGetOCRResult,
+                     weak_ptr_factory_.GetWeakPtr()),
+      active_chat_tab_helper_->GetMaxPageContentLength(),
+      pdf_use_skia_renderer_enabled);
+  preview_page_text_extractor_->StartExtract();
+}
+
+void AIChatUIPageHandler::OnGetOCRResult(std::string text) {
+  active_chat_tab_helper_->OnPreviewTextReady(std::move(text));
+  PreviewCleanup();
+}
+
+void AIChatUIPageHandler::MaybeCreatePrintPreview() {
+  auto url = active_chat_tab_helper_->web_contents()->GetLastCommittedURL();
+  if (!base::Contains(kPrintPreviewRetrievalHosts, url.host_piece())) {
+    return;
+  }
+  // TODO(darkdh): support pdf preview printing
+  content::RenderFrameHost* rfh =
+      printing::GetFrameToPrint(active_chat_tab_helper_->web_contents());
+  if (rfh) {
+    if (!print_render_frame_.is_bound()) {
+      rfh->GetRemoteAssociatedInterfaces()->GetInterface(&print_render_frame_);
+    }
+
+    print_render_frame_->InitiatePrintPreview(false);
```

Comment:
there's `DisablePrintPreview` group policy. Please make sure things don't explode when it's enabled.

it's possible the content extraction should be inaccessible if the feature cannot operate.

Reply:
It's being handled in 7bdce8d69d66ca65baf9ea937e06f1cab48c828e